atc/atccmd/command.go

```diff
@@ -1024,46 +995,18 @@ func (cmd *RunCommand) backendComponents(
 	dbJobFactory := db.NewJobFactory(dbConn, lockFactory)
 	dbPipelineLifecycle := db.NewPipelineLifecycle(dbConn, lockFactory)
 
-	alg := algorithm.New(db.NewVersionsDB(dbConn, algorithmLimitRows, schedulerCache))
-
-	dbWorkerBaseResourceTypeFactory := db.NewWorkerBaseResourceTypeFactory(dbConn)
-	dbTaskCacheFactory := db.NewTaskCacheFactory(dbConn)
-	dbWorkerTaskCacheFactory := db.NewWorkerTaskCacheFactory(dbConn)
 	dbVolumeRepository := db.NewVolumeRepository(dbConn)
+	resourceGetter := resource.NewGetter(lockFactory, clock.NewClock(), dbResourceCacheFactory, dbVolumeRepository)
```

Comment:
Agreed that `resource.Getter` doesn't really have a strong identity, but not sure I agree with turning it into a `ResourceFactory`/`ContainerFactory`. `resource.Getter` is really just responsible for executing a part of the get step - specifically, looking up the resource cache, and if it doesn't exist, running the `in` script on resources under a lock and storing it as a cache. It doesn't know about how to create the container nor how to create the resource (which is just data). It also doesn't *really* know how to run the `in` script, and can easily be modified to support prototype-based resources by just taking an interface rather than the concrete `resource.Resource`.

IMO the better approach would be to move this logic into either the `exec.GetStep` itself or `engine.GetStepDelegate`, or some combination thereof (letting the delegate handle the DB interaction). The main reason for not doing this initially was so that I could test this logic in isolation, but it may not actually be that bad testing at the level of the get step.

Reply:
Take a look at https://github.com/concourse/concourse/pull/6597/commits/ab30714be27030b68b857b62740066fa51f316ca, which implements this idea. Moving the `resource.Getter` into `exec.GetStep` also allowed for simplifying some redundant cache-checking logic that was added for caching streamed volumes.

p.s. sorry for adding yet another big commit into this already massive PR